File: cloneGraph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;

    Node() {}

    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
public:
    map<Node*,Node*> dic;
    Node* cloneGraph(Node* node) {
        if(!node) return nullptr;
        if(dic.count(node)) return dic[node];
        Node* cur = new Node(node->val);
        dic[node] = cur;
        for(int i = 0;i<node->neighbors.size();i++){
            cur->neighbors.push_back(cloneGraph(node->neighbors[i]));
        }
        return cur;
    }
};

File: cloneGraph.py
class Solution:
    # dfs的递归做法，憨批题目
    def cloneGraph(self, node: 'Node') -> 'Node':
        lookup = {}
        def dfs(node):
            if not node:
                return
            if node in lookup:
                return lookup[node]
            clone = Node(node.val,[])
            lookup[node] = clone
            for n in node.neighbors:
                clone.neighbors.append(dfs(n))
            return clone
        return dfs(node)
